### PR TITLE
Add subfield support to DomainTranslator

### PR DESCRIPTION
**Pull request by mbasmanova**
Modify DomainTranslator to support subfields, e.g. `a[2] > 10`.

Part of #12585

### CODE REVIEW

File: presto-spi/src/test/java/com/facebook/presto/spi/TestSubfieldTokenizer.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        List<Subfield.PathElement> elements = ImmutableList.of(
+                new NestedField("b"),
+                new Subfield.LongSubscript(2),
+                new Subfield.StringSubscript("z"),
+                Subfield.allSubscripts(),
+                new Subfield.StringSubscript("34"),
+                new Subfield.StringSubscript("b \"test\""),
+                new Subfield.StringSubscript("\"abc"),
+                new Subfield.StringSubscript("abc\""),
+                new Subfield.StringSubscript("ab\"cde"));
```
Comment (highker) [inline presto-spi/src/test/java/com/facebook/presto/spi/TestSubfieldTokenizer.java]: I think it worth adding some strange UTF-8 names. Corner cases are emoji, invalid UTF-8, or (valid or invalid) surrogates. For example `"a.b[\"hello\uDBFF\"]"`. There are many examples in `TestStringFunctions`.

File: presto-spi/src/main/java/com/facebook/presto/spi/SubfieldTokenizer.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        int end = index;
+
+        String token = path.substring(start, end);
+
+        // an empty unquoted token is not allowed
+        if (token.isEmpty()) {
+            throw invalidSubfieldPath();
+        }
+
+        return new Subfield.LongSubscript(Long.valueOf(token));
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/SubfieldTokenizer.java]: We might need to check if `token` is a legit long. For example `a[b]`

File: presto-spi/src/test/java/com/facebook/presto/spi/TestSubfieldTokenizer.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestSubfieldTokenizer
+{
+    @Test
+    public void test()
```
Comment (highker) [inline presto-spi/src/test/java/com/facebook/presto/spi/TestSubfieldTokenizer.java]: Maybe have another test to cover some illegal cases.

File: presto-spi/src/main/java/com/facebook/presto/spi/Subfield.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        checkArgument(firstElement instanceof NestedField, "Subfield path must start with a column name: " + path);
+
+        this.name = ((NestedField) firstElement).getName();
+
+        List<PathElement> pathElements = new ArrayList<>();
+        tokenizer.forEachRemaining(pathElements::add);
+        this.path = Collections.unmodifiableList(pathElements);
+    }
+
+    private static void checkArgument(boolean expression, String errorMessage)
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/Subfield.java]: This is essentially a copy of a method from Guava. The parameter names match the original.

**Comment by mbasmanova** [comment]
@highker James, thank you for review. I'll address your comments and update the PR shortly.

> Subfield together with its extractor does not need to live in SPI. 

I need Subfield to add an optimizer rule that collects all the referenced subfields and provides these to the connector for pruning. For example, the following query needs only one field from a struct. 

`SELECT a.b FROM t`

File: presto-spi/src/main/java/com/facebook/presto/spi/Subfield.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class Subfield
+{
+    public abstract static class PathElement
```
Comment (tdcmeehan) [inline presto-spi/src/main/java/com/facebook/presto/spi/Subfield.java]: Is it abstract for performance reasons?

Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/Subfield.java]: Actually, no. There is no reason for it to be an abstract class. I'll change it to an interface.

**Comment by mbasmanova** [comment]
@tdcmeehan @highker James, Tim, thank you for reviewing. All comments addressed. Also, I added a separate test for SubfieldExtractor and adding some more predicates to TestDomainTranslator in presto-hive.

**Pull request merged**